_posts/databases/postgresql/timescaledb/2000-01-01-downsampling.md

```diff
@@ -0,0 +1,122 @@
+
+Now we want to decrease the database size. We don't need to have data by the
+minute for data older than one week, the average of values by hour is enough.
+So we can use a downsampling logic which will reduce the datas older than one week,
+and only store the average by hour.
+
+
+### Datas
+
+For the following table:
+
+```
+$ select * from conditions;
+             time              | location |    temperature     |      humidity
+-------------------------------+----------+--------------------+--------------------
+ 2022-01-21 09:50:05.049911+00 | office   |                 75 |                 52
+ 2022-01-21 09:50:05.058412+00 | office   |                 70 |                 57
+ 2022-01-21 09:50:05.059489+00 | office   |                 76 |                 52
+ 2022-01-20 09:50:05.06032+00  | office   |                 79 |                 54
+```
+
+### Downsample Procedure
+
+The procedure bellow contains the logic to downsample the table `conditions`
+
+```sql
+CREATE OR REPLACE PROCEDURE downsample_conditions (config jsonb)
+LANGUAGE PLPGSQL
+AS $$
+DECLARE
+  lag interval;
+  chunk REGCLASS;
+  tmp_name name;
+BEGIN
+  SELECT jsonb_object_field_text (config, 'lag')::interval INTO STRICT lag;
+
+  IF lag IS NULL THEN
+    RAISE EXCEPTION 'Config must have lag';
+  END IF;
+
+  FOR chunk IN
+    SELECT show.oid
+    -- we specify the table which need to be downsampled
+    FROM show_chunks('conditions', older_than => lag) SHOW (oid)
+      INNER JOIN pg_class pgc ON pgc.oid = show.oid
+      INNER JOIN pg_namespace pgns ON pgc.relnamespace = pgns.oid
+      INNER JOIN timescaledb_information.chunks chunk ON chunk.chunk_name = pgc.relname
+        AND chunk.chunk_schema = pgns.nspname
+  LOOP
+    RAISE NOTICE 'Processing chunk: %', chunk::text;
+
+    -- build name for temp table
+    SELECT '_tmp' || relname
+    FROM pg_class
+    WHERE oid = chunk INTO STRICT tmp_name;
+
+    -- copy downsampled chunk data into temp table
+    EXECUTE format($sql$ CREATE UNLOGGED TABLE %I AS
+      -- you can configure here the time range to aggregate datas and how you do it. Here we used `avg` function on 1 hour
+      SELECT time_bucket('1h', time), location, avg(temperature) as avg_temperature, avg(humidity) as avg_humidity FROM %s GROUP BY 1, 2;
+    $sql$, tmp_name, chunk);
+
+    -- clear original chunk
+    EXECUTE format('TRUNCATE %s;', chunk);
+
+    -- copy downsampled data back into chunk
+    EXECUTE format('INSERT INTO %s(time, location, temperature, humidity) SELECT * FROM %I;', chunk, tmp_name);
+
+    -- drop temp table
+    EXECUTE format('DROP TABLE %I;', tmp_name);
+
+    COMMIT;
+  END LOOP;
+END
+$$;
+```
+
+### Execute Procedure
+
+You can schedule the execution of the previous procedure by using
+[Scalingo scheduler]({% post_url platform/app/task-scheduling/2000-01-01-scalingo-scheduler %}).
+To do so, add a `cron.json` file at the root of your application’s source code,
```

Comment:
nitpick: not sure `To do so` adds anything here. And it really sounds like a French translation to my hears. I might be wrong:
```suggestion
Add a `cron.json` file at the root of your application’s source code,
```